prez/routers/sparql.py

```diff
@@ -36,16 +39,104 @@ async def sparql(
+            # the user cannot know which endpoint each result came from.
+            # no current ideas for a better solution.
+            elif results[0][1] == ("text/turtle"):
+                graph = Graph(bind_namespaces="rdflib")
+                graph.bind("prez", "https://prez.dev/")
+                for res in results:
+                    g = res[2]
+                    graph.__iadd__(g)
+                return await return_rdf(
+                    graph, mediatype="text/anot+turtle", profile_headers={}
+                )
+    else:
+        prez = pathmap.get(start_of_path)
+    _, mt, result = await sparql_query(query, prez)
+    if mt.startswith("application/json"):
+        return JSONResponse(result)
+    if len(result) > 0:
+        return await return_rdf(result, mediatype=mt, profile_headers={})
+    else:
+        return PlainTextResponse(f"No results returned from {prez} SPARQL endpoint.")
+
+
+# class Query(BaseModel):
+#     q: str
+
+
+@router.post("/sparql", summary="Federated Sparql")
+@router.post("/v/sparql", summary="VocPrez Sparql")
+@router.post("/c/sparql", summary="CatPrez Sparql")
+@router.post("/s/sparql", summary="SpacePrez Sparql")
+async def sparql(
+    request: Request, query: str = Body(..., media_type="application/sparql-query")
+):
+    from prez.config import settings
+
+    if not query:
+        query = request.query_params.get("query")
+        if not query:
             return PlainTextResponse(
-                "Federated SPARQL endpoint not yet implemented. Use the individual SPARQL "
-                "endpoints: /v/sparql, /s/sparql, and /c/sparql"
+                "A query must be provided as a query string argument (?query=<SPARQL query>)"
+            )
+    start_of_path = request.url.path[:3]
+    pathmap = {
+        "/v/": "VocPrez",
+        "/c/": "CatPrez",
+        "/s/": "SpacePrez",
+    }
+    if start_of_path == "/sp":
+        if len(settings.enabled_prezs) == 1:
+            prez = settings.enabled_prezs[0]
+        else:
+            results = await asyncio.gather(
+                *[sparql_query(query, p) for p in settings.enabled_prezs]
             )
+            for result in results:
+                if isinstance(result, PlainTextResponse):
+                    return result
+            if results[0][1].startswith("application/json"):
+                combined_bindings = []
+                for result in results:
+                    for binding in result[2]["results"]["bindings"]:
+                        binding["endpoint"] = {"type": "uri", "value": result[0]}
+                        combined_bindings.append(binding)
+                results[0][2]["results"]["bindings"] = combined_bindings
+                return JSONResponse(results[0][2])
+            # warning: the following is not a sensible way to return results.
+            # the user cannot know which endpoint each result came from.
+            # no current ideas for a better solution.
```

Comment:
We can inject a new SPARQL response var and column for each row since we know which endpoint returned the result when we query it.

Example: we inject a new var and column for each row named `system:prez:endpoint` to indicate which endpoint the results came from.

| s   | p   | o   | system:prez:endpoint                 |
|-----|-----|-----|--------------------------------------|
| `<a>` | `<b>` | `<c>` | https://example.com/vocprez/sparql   |
| `<d>` | `<e>` | `<f>` | https://example.com/catprez/sparql   |
| `<g>` | `<h>` | `<i>` | https://example.com/spaceprez/sparql |

Reply:
To be clear it's where an RDF mediatype is returned that I haven't added a way to know which endpoint the results came from.

Where the results are tabular application/sparql-results (picked up with the "application/json" mediatype - need to check why application/sparql-results isn't the MT) I am adding the endpoint in as an additional binding via "endpoint"